File: src/main/java/com/vectordb/DocumentDB.java
package com.vectordb;

import java.util.*;
import java.util.concurrent.locks.ReentrantLock;

/**
 * DocumentDB — HNSW index over real Ollama embeddings (768D).
 * Handles the RAG document storage with HNSW + brute force fallback.
 */
public class DocumentDB {

    public static class DocItem {
        public final int id;
        public final String title;
        public final String text;
        public final float[] emb;

        public DocItem(int id, String title, String text, float[] emb) {
            this.id = id;
            this.title = title;
            this.text = text;
            this.emb = emb;
        }
    }

    private final Map<Integer, DocItem> store = new HashMap<>();
    private final HNSW hnsw;
    private final BruteForce bf = new BruteForce();
    private final ReentrantLock lock = new ReentrantLock();
    private int nextId = 1;
    private int dims = 0;

    public DocumentDB() {
        this.hnsw = new HNSW(16, 200);
    }

    /**
     * Insert a pre-embedded document chunk.
     */
    public int insert(String title, String text, float[] emb) {
        lock.lock();
        try {
            if (dims == 0) dims = emb.length;
            DocItem item = new DocItem(nextId++, title, text, emb);
            store.put(item.id, item);

            VectorItem vi = new VectorItem(item.id, title, "doc", emb);
            hnsw.insert(vi, DistanceMetrics::cosine);
            bf.insert(vi);

            return item.id;
        } finally {
            lock.unlock();
        }
    }

    /**
     * Semantic search — returns top-k most similar document chunks
     * with their distances, filtered by max_dist threshold.
     */
    public List<Map.Entry<Float, DocItem>> search(float[] query, int k, float maxDist) {
        lock.lock();
        try {
            if (store.isEmpty()) return Collections.emptyList();

            List<float[]> raw;
            if (store.size() < 10) {
                raw = bf.knn(query, k, DistanceMetrics::cosine);
            } else {
                raw = hnsw.knn(query, k, 50, DistanceMetrics::cosine);
            }

            List<Map.Entry<Float, DocItem>> results = new ArrayList<>();
            for (float[] pair : raw) {
                int id = (int) pair[1];
                float dist = pair[0];
                DocItem doc = store.get(id);
                if (doc != null && dist <= maxDist) {
                    results.add(Map.entry(dist, doc));
                }
            }
            return results;
        } finally {
            lock.unlock();
        }
    }

    public List<Map.Entry<Float, DocItem>> search(float[] query, int k) {
        return search(query, k, 0.7f);
    }

    public boolean remove(int id) {
        lock.lock();
        try {
            if (!store.containsKey(id)) return false;
            store.remove(id);
            hnsw.remove(id);
            bf.remove(id);
            return true;
        } finally {
            lock.unlock();
        }
    }

    public List<DocItem> all() {
        lock.lock();
        try {
            return new ArrayList<>(store.values());
        } finally {
            lock.unlock();
        }
    }

    public int size() {
        lock.lock();
        try {
            return store.size();
        } finally {
            lock.unlock();
        }
    }

    public int getDims() {
        return dims;
    }
}

File: vectordb/document_db.py
"""
document_db.py — HNSW index over real Ollama embeddings (768D).

Handles the RAG document storage: stores document chunks with their 768D
embeddings and supports similarity search for the RAG pipeline.
Uses HNSW for large datasets and falls back to BruteForce for < 10 chunks.

Thread-safe via threading.Lock.
Equivalent to DocumentDB.java in the original Java project.
"""

import threading
from typing import Dict, List, Optional, Tuple

from .brute_force import BruteForce
from .distance_metrics import cosine, get_dist_fn
from .hnsw import HNSW
from .models import DocItem, VectorItem


class DocumentDB:
    """
    HNSW-backed store for document chunks with 768D Ollama embeddings.

    insert()   — add a chunk (pre-embedded)
    remove()   — remove a chunk by id
    search()   — return top-k similar chunks
    list_all() — return all stored chunks
    """

    def __init__(self) -> None:
        self._store: Dict[int, DocItem] = {}
        self._hnsw = HNSW(M=16, ef_build=200)
        self._bf = BruteForce()
        self._lock = threading.Lock()
        self._next_id = 1
        self._dims = 0

    def insert(self, title: str, text: str, emb: List[float]) -> int:
        """Insert a pre-embedded document chunk. Returns the assigned ID."""
        dist_fn = cosine
        with self._lock:
            item_id = self._next_id
            self._next_id += 1
            self._dims = len(emb)

            doc = DocItem(id=item_id, title=title, text=text, emb=emb)
            self._store[item_id] = doc

            # Build a VectorItem wrapper for the generic indexes
            vi = VectorItem(id=item_id, metadata=title, category="doc", emb=emb)
            self._bf.insert(vi)
            self._hnsw.insert(vi, dist_fn)
            return item_id

    def remove(self, item_id: int) -> bool:
        """Remove a document chunk by ID."""
        with self._lock:
            if item_id not in self._store:
                return False
            del self._store[item_id]
            self._bf.remove(item_id)
            self._hnsw.remove(item_id)
            return True

    def search(self, query_emb: List[float], k: int) -> List[DocItem]:
        """
        Return the top-k most similar document chunks for the given embedding.
        Uses HNSW for ≥10 chunks, brute force otherwise.
        """
        dist_fn = cosine
        with self._lock:
            if not self._store:
                return []
            if len(self._store) < 10:
                raw = self._bf.knn(query_emb, k, dist_fn)
            else:
                raw = self._hnsw.knn(query_emb, k, dist_fn)

            results: List[DocItem] = []
            for _dist, nid in raw:
                doc = self._store.get(nid)
                if doc is not None:
                    results.append(doc)
            return results

    def list_all(self) -> List[DocItem]:
        with self._lock:
            return list(self._store.values())

    def size(self) -> int:
        with self._lock:
            return len(self._store)
